controllers/s3utils.go

```diff
@@ -0,0 +1,386 @@
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Example usage:
+// func example_code() {
+// *** setup a new s3 object store ***
+// s3endpoint := "http://127.0.0.1:9000"
+// s3secretname := types.namespacedname{name: s3secretname, namespace: parent.namespace}
+
+// s3conn, err := connecttos3endpoint(ctx, reconciler, s3endpoint, s3secretname)
+// if err != nil {
+// 	return err
+// }
+// *** create a new bucket ***
+// bucket := "subname-namespace" // should be all lowercase
+// if err := s3Conn.createBucket(bucket); err != nil {
+// 	return err
+// }
+
+// *** Upload objects, optionally using a key prefix to easily find the objects later ***
+// for i := 1; i < 10; i++ {
+// 	pvKey := fmt.Sprintf("PersistentVolumes/pv%v", i)
+// 	uploadPV := corev1.PersistentVolume{}
+// 	uploadPV.Name = pvKey
+// 	uploadPV.Spec.StorageClassName = "gold"
+// 	uploadPV.Spec.PersistentVolumeReclaimPolicy = corev1.PersistentVolumeReclaimRetain
+// 	if err := s3Conn.uploadObject(bucket, pvKey, uploadPV); err != nil {
+// 		return err
+// 	}
+// }
+
+// *** Find objects in the bucket, optionally supplying a key prefix
+// keyPrefix := "v1.PersistentVolumes/"
+// if list, err := s3Conn.listKeys(bucket, keyPrefix); err != nil {
+// 	return err
+// } else {
+// 	for _, key := range list {
+// 		fmt.Printf("%v ", key)
+// 	}
+// }
+
+// *** Download from the given bucket an object with the given key
+// keyPrefix := "v1.PersistentVolumes/"
+// key := keyPrefix + "pv2"
+// var downloadPV corev1.PersistentVolume
+// if err := s3Conn.downloadObject(bucket, key, &downloadPV); err != nil {
+// 	return err
+// }
+// }
+
+type s3Connection struct {
+	session    *session.Session
+	client     *s3.S3
+	uploader   *s3manager.Uploader
+	downloader *s3manager.Downloader
+	endpoint   string
+	callerTag  string
+}
+
+var objectStoreConnections = map[string]*s3Connection{}
+
+// Return a connection to the given S3 object store endpoint using
+// the given s3 secret
+// - Return error if endpoint or secret is not configured, or if
+//   session creation fails
+func connectToS3Endpoint(ctx context.Context, r client.Reader,
+	s3Endpoint string, s3SecretName types.NamespacedName,
+	callerTag string) (*s3Connection, error) {
+	if s3Endpoint == "" {
+		return nil, fmt.Errorf("s3 endpoint has not been configured; tag:%s",
+			callerTag)
+	}
+
+	// Use cached connection, if one exists
+	if s3Conn, ok := objectStoreConnections[s3Endpoint]; ok {
+		return s3Conn, nil
+	}
+
+	accessID, secretAccessKey, err := getS3Secret(ctx, r, s3SecretName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get secret %v; tag %s, %w",
+			s3SecretName, callerTag, err)
+	}
+
+	// Create an S3 client session
+	s3Session, err := session.NewSession(&aws.Config{
+		Credentials: credentials.NewStaticCredentials(string(accessID),
+			string(secretAccessKey), ""),
+		Endpoint:         aws.String(s3Endpoint),
+		Region:           aws.String("us-east-1"),
```

Comment:
(nit) Shoud the region be something that has to be populated based on some input provided?

Reply:
When using an S3 store like velero, the concept of region doesn't apply and default's to us-east-1.  If we need to support a cloud based S3 store, yes, we may need take region as an input, among other TBD parameters. Good point.

Reply:
I created https://github.com/RamenDR/ramen/issues/36 to track this issue.